core/src/main/scala/kafka/log/Log.scala

```diff
@@ -470,22 +470,44 @@ class Log(val dir: File,
     trace("Reading %d bytes from offset %d in log %s of length %d bytes".format(maxLength, startOffset, name, size))
 
-    // check if the offset is valid and in range
-    val next = nextOffsetMetadata.messageOffset
-    if(startOffset == next)
-      return FetchDataInfo(nextOffsetMetadata, MessageSet.Empty)
-    
+    // We first create a few local variables to avoid race conditions with updates to the log. The order matters here.
+    // We can allow currentLastEntry to have a newer value than currentNextOffsetMetadata but not vice versa.
+    val currentNextOffsetMetadata = nextOffsetMetadata
+    val currentNext = currentNextOffsetMetadata.messageOffset
+    val currentLastEntry = segments.lastEntry
+
+    if(startOffset == currentNext)
+      return FetchDataInfo(currentNextOffsetMetadata, MessageSet.Empty)
+
     var entry = segments.floorEntry(startOffset)
-      
+
     // attempt to read beyond the log end offset is an error
-    if(startOffset > next || entry == null)
-      throw new OffsetOutOfRangeException("Request for offset %d but we only have log segments in the range %d to %d.".format(startOffset, segments.firstKey, next))
+    if(startOffset > currentNext || entry == null)
+      throw new OffsetOutOfRangeException(("Request for offset %d but we only have log segments in the range %d " +
+        "to %d.").format(startOffset, segments.firstKey, currentNext))
     
-    // do the read on the segment with a base offset less than the target offset
+    // Do the read on the segment with a base offset less than the target offset
     // but if that segment doesn't contain any messages with an offset greater than that
     // continue to read from successive segments until we get some messages or we reach the end of the log
+    // Note that at this point startOffset < currentNext. There are three cases:
+    // 1. No new log segment is created after we created local variable currentNextOffsetMetadata and currentLastEntry
+    // 2. A new log segment is created after we created local variable currentNextOffsetMetadata but before we created
+    //    currentLastEntry. i.e currentLastEntry is actually newer than currentNextOffsetMetadata
+    // 3. A new log segment is created after we created currentNextOffsetMetadata and currentLastEntry.
+    // In all three cases, reading will start from currentLastEntry or older segments, but not the newly rolled segment.
+    // In case (2) and (3), if the reading starts from the currentLastEntry, it might be reading a little conservatively
+    // because we capped the reading up to currentNextOffsetMetadata.relativePositionInSegment even though the log
+    // segment being read is no longer the active log segment. But it is fine because later reading will see updated
+    // messages.
     while(entry != null) {
-      val fetchInfo = entry.getValue.read(startOffset, maxOffset, maxLength)
+      val maxPosition = {
+        if (entry == currentLastEntry) {
```

Comment:
Thinking a bit more about this logic. It seems that we can still have a potential problem, in theory. For example, in a compacted topic, let's say we run through line 480 in read(). At this moment, currentLastEntry is L1 and startOffset is in segment L1. Then we add a new segment L2 and also do the log cleaning, which removes all messages in L1 after startOffset and replaces L1 with a new cleaned segment L1'. After this, read() continues from line 481. When going through the while loop in line 502, in the first iteration, we will try to read from segment L1' and it will return null since there is no message after startOffset. Then we go to the next segment, which is L2. Since L2 is not currentLastEntry, we will be using entry.getValue.size. Now we will run the risk of exposing data after nextOffset.